File: apps/windows/TeslaSync.App.Core/Notifications/NotificationRouteMap.cs
using TeslaSync.App.Core.Navigation;

namespace TeslaSync.App.Core.Notifications;

/// <summary>The deep-link target a notification opens: an in-app route path plus the entity id it was built from.</summary>
public sealed record ResolvedRoute(string Path, string? EntityId);

/// <summary>
/// Maps a <see cref="NotificationKind"/> (and the push <c>data</c> bag) to a concrete in-app route
/// path (P2/W8-0001). Every candidate is validated against the real <see cref="RouteRegistry"/> — the
/// same registry the W3 shell navigates — so a toast or jump-list entry can never deep-link to a route
/// that does not exist; an unresolvable kind falls back to the always-present notifications inbox.
///
/// <para>Resolution order: (1) an explicit, valid <c>route</c> the backend supplied wins; (2) the
/// kind's parameterized candidate (with a safe entity id) then its static landing page; (3) the inbox.</para>
/// </summary>
public static class NotificationRouteMap
{
    /// <summary>The canonical, always-valid fallback route (the notifications inbox).</summary>
    public const string InboxPath = "notifications/inbox";

    private static readonly string[] VehicleIdKeys = { "vehicle_id", "vehicleId", "id" };
    private static readonly string[] SessionIdKeys = { "session_id", "sessionId", "charging_session_id", "chargingSessionId", "id" };
    private static readonly string[] IncidentIdKeys = { "incident_id", "incidentId", "id" };
    private static readonly string[] GenericIdKeys = { "id", "vehicle_id", "vehicleId" };

    /// <summary>Resolves the deep-link route for <paramref name="kind"/> given its push <paramref name="data"/>.</summary>
    public static ResolvedRoute Resolve(
        NotificationKind kind,
        IReadOnlyDictionary<string, string> data,
        RouteRegistry registry)
    {
        ArgumentNullException.ThrowIfNull(data);
        ArgumentNullException.ThrowIfNull(registry);

        var entityId = EntityIdFor(kind, data);

        if (data.TryGetValue(ToastArguments.RouteKey, out var explicitRoute) && IsReal(registry, explicitRoute))
        {
            return new ResolvedRoute(RouteRegistry.Normalize(explicitRoute), entityId);
        }

        foreach (var candidate in CandidatesFor(kind, entityId))
        {
            if (IsReal(registry, candidate))
            {
                return new ResolvedRoute(RouteRegistry.Normalize(candidate), entityId);
            }
        }

        return new ResolvedRoute(InboxPath, entityId);
    }

    private static IEnumerable<string> CandidatesFor(NotificationKind kind, string? entityId)
    {
        switch (kind)
        {
            case NotificationKind.Alert:
                yield return "notifications/alerts";
                yield return InboxPath;
                break;
            case NotificationKind.ChargeComplete:
                if (entityId is not null)
                {
                    yield return "charging/" + entityId;
                }

                yield return "charging";
                break;
            case NotificationKind.VehicleState:
                if (entityId is not null)
                {
                    yield return "vehicles/" + entityId;
                }

                yield return "vehicles";
                break;
            case NotificationKind.Automation:
                yield return "automations";
                break;
            case NotificationKind.CommandResult:
                yield return "command-history";
                yield return "commands";
                break;
            case NotificationKind.SystemIncident:
                if (entityId is not null)
                {
                    yield return "system-status/incidents/" + entityId;
                }

                yield return "system-status";
                break;
            case NotificationKind.ReauthNeeded:
                yield return "settings";
                break;
            default:
                yield return InboxPath;
                break;
        }
    }

    private static string? EntityIdFor(NotificationKind kind, IReadOnlyDictionary<string, string> data)
    {
        var keys = kind switch
        {
            NotificationKind.ChargeComplete => SessionIdKeys,
            NotificationKind.VehicleState => VehicleIdKeys,
            NotificationKind.SystemIncident => IncidentIdKeys,
            _ => GenericIdKeys,
        };

        foreach (var key in keys)
        {
            if (data.TryGetValue(key, out var value) && IsSafeSegment(value))
            {
                return value.Trim();
            }
        }

        return null;
    }

    private static bool IsReal(RouteRegistry registry, string? path) =>
        !string.IsNullOrWhiteSpace(path) && !registry.Resolve(path).Route.IsCatchAll;

    private static bool IsSafeSegment(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        foreach (var c in value.Trim())
        {
            if (!char.IsLetterOrDigit(c) && c != '-' && c != '_' && c != '.')
            {
                return false;
            }
        }

        return true;
    }
}

File: apps/android/app/src/main/java/io/teslasync/android/notifications/NotificationRouteMap.kt
package io.teslasync.android.notifications

import io.teslasync.android.navigation.RouteTable

/** The deep-link target a notification opens: an in-app route path plus the entity id it was built from. */
data class ResolvedRoute(
    val path: String,
    val entityId: String?,
)

/**
 * Maps a [NotificationKind] (and the push `data` bag) to a concrete in-app route path and a
 * `teslasync://app/...` deep-link URI (P3/A6, ADR-009). Every candidate is validated against the real
 * [RouteTable] — the same registry the navigation shell uses — so a notification can never deep-link to
 * a route that does not exist; an unresolvable kind falls back to the always-present notifications inbox.
 *
 * Resolution order: (1) an explicit, valid `route` the backend supplied wins; (2) the kind's
 * parameterized candidate (with a safe entity id) then its static landing page; (3) the inbox.
 */
object NotificationRouteMap {
    /** The canonical, always-valid fallback route (the notifications inbox). */
    const val INBOX_PATH = "notifications/inbox"

    private val vehicleIdKeys = listOf("vehicle_id", "vehicleId", "id")
    private val sessionIdKeys = listOf("session_id", "sessionId", "charging_session_id", "chargingSessionId", "id")
    private val incidentIdKeys = listOf("incident_id", "incidentId", "id")
    private val genericIdKeys = listOf("id", "vehicle_id", "vehicleId")

    /** Resolves the deep-link route for [kind] given its push [data]. */
    fun resolve(
        kind: NotificationKind,
        data: Map<String, String>,
    ): ResolvedRoute {
        val entityId = entityIdFor(kind, data)
        val explicit = data["route"]
        val resolved =
            when {
                explicit != null && isReal(explicit) -> normalize(explicit)
                else -> firstRealCandidate(kind, entityId) ?: INBOX_PATH
            }
        return ResolvedRoute(resolved, entityId)
    }

    /** The full `teslasync://app/...` deep-link URI a tap on a [kind] notification should open. */
    fun deepLinkUri(
        kind: NotificationKind,
        data: Map<String, String>,
    ): String = deepLinkUriFor(resolve(kind, data).path)

    /** Builds the `teslasync://app/...` deep-link URI for an already-resolved in-app [path]. */
    fun deepLinkUriFor(path: String): String = "${RouteTable.APP_SCHEME}://app/" + path.removePrefix("/")

    private fun firstRealCandidate(
        kind: NotificationKind,
        entityId: String?,
    ): String? = candidatesFor(kind, entityId).firstOrNull { isReal(it) }?.let { normalize(it) }

    private fun candidatesFor(
        kind: NotificationKind,
        entityId: String?,
    ): List<String> =
        when (kind) {
            NotificationKind.Alert -> listOf("notifications/alerts", INBOX_PATH)
            NotificationKind.ChargeComplete ->
                buildList {
                    if (entityId != null) add("charging/$entityId")
                    add("charging")
                }
            NotificationKind.VehicleState ->
                buildList {
                    if (entityId != null) add("vehicles/$entityId")
                    add("vehicles")
                }
            NotificationKind.Automation -> listOf("automations")
            NotificationKind.CommandResult -> listOf("command-history", "commands")
            NotificationKind.SystemIncident ->
                buildList {
                    if (entityId != null) add("system-status/incidents/$entityId")
                    add("system-status")
                }
            NotificationKind.ReauthNeeded -> listOf("settings")
            NotificationKind.Generic -> listOf(INBOX_PATH)
        }

    private fun entityIdFor(
        kind: NotificationKind,
        data: Map<String, String>,
    ): String? {
        val keys =
            when (kind) {
                NotificationKind.ChargeComplete -> sessionIdKeys
                NotificationKind.VehicleState -> vehicleIdKeys
                NotificationKind.SystemIncident -> incidentIdKeys
                else -> genericIdKeys
            }
        return keys.firstNotNullOfOrNull { key -> data[key]?.takeIf(::isSafeSegment)?.trim() }
    }

    private fun isReal(path: String): Boolean {
        val destination = RouteTable.match(path)
        return destination != null && destination.id != RouteTable.notFound.id
    }

    private fun normalize(path: String): String = RouteTable.normalize(path).removePrefix("/")

    private fun isSafeSegment(value: String?): Boolean {
        if (value.isNullOrBlank()) return false
        return value.trim().all { char -> char.isLetterOrDigit() || char == '-' || char == '_' || char == '.' }
    }
}
